File: 1510-stone-game-iv/1510-stone-game-iv.py
class Solution:
    def winnerSquareGame(self, n: int) -> bool:
        # dp[i] represents if the current player can force a win with 'i' stones
        dp = [False] * (n + 1)
        
        dp[0] = False
        
        for i in range(1, n + 1):
            j = 1
            while j * j <= i:
                square = j * j
                
                # If the remaining stones put the opponent in a losing state, we win
                if not dp[i - square]:
                    dp[i] = True
                    break
                j += 1
                
        return dp[n]

File: 1510-stone-game-iv/1510-stone-game-iv.js
/**
 * @param {number} n
 * @return {boolean}
 */
var winnerSquareGame = function(n) {
    // Array allocation for DP state caching
    const dp = new Array(n + 1).fill(false);
    
    dp[0] = false;
    
    for (let i = 1; i <= n; i++) {
        for (let j = 1; j * j <= i; j++) {
            const square = j * j;
            
            if (!dp[i - square]) {
                dp[i] = true;
                break;
            }
        }
    }
    
    return dp[n];
};
